[NFC][lldb] Adapt LLDB to RemoteInspection's RemoteAddress changes

augusto2112 opened the pull request:
Adapts LLDB to the new field in RemoteAddress. This is an NFC that
maintains the current behavior. A follow up patch will be introduced
later that takes advantage of the new behavior to fix a bug in reading
metadata from files instead of the process.

rdar://148361743

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci  test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1053,7 +1053,6 @@ void SwiftASTContext::SetCompilerInvocationLLDBOverrides() {
   // When loading Swift types that conform to ObjC protocols that have
   // been renamed with NS_SWIFT_NAME the DwarfImporterDelegate will crash
   // during protocol conformance checks as the underlying type cannot be
-  // found. Allowing module compilation to proceed with compiler
```
looks like this is an unintentional removal

kastiglione commented on lldb/source/Plugins/TypeSystem/Swift/SwiftASTContext.cpp:
```diff
@@ -1080,19 +1079,15 @@ void SwiftASTContext::SetCompilerInvocationLLDBOverrides() {
 }
 
 SwiftASTContext::~SwiftASTContext() {
-#ifndef NDEBUG
-  m_ast_context_mutex.lock();
   if (swift::ASTContext *ctx = m_ast_context_ap.get())
     // A RemoteASTContext associated with this swift::ASTContext has
     // to be destroyed before the swift::ASTContext is destroyed.
     assert(!GetASTMap().Lookup(ctx) && "ast context still in global map");
-  m_ast_context_mutex.unlock();
-#endif
 }
 
 SwiftASTContextForModule::~SwiftASTContextForModule() {
-  if (auto ctx = GetASTContext())
-    GetASTMap().Erase(*ctx);
```
these don't seem clearly related, what motivated these?

augusto2112 replied:
No idea how this happened.

kastiglione commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftLanguageRuntimeRemoteAST.cpp:
```diff
@@ -119,7 +119,8 @@ std::optional<uint64_t> SwiftLanguageRuntime::GetMemberVariableOffsetRemoteAST(
       if (!pointer || pointer == LLDB_INVALID_ADDRESS)
         break;
-      swift::remote::RemoteAddress address(pointer);
+      auto address = swift::remote::RemoteAddress(
+          pointer, swift::remote::RemoteAddress::DefaultAddressSpace);
```
could the second parameter have a default value of `swift::remote::RemoteAddress::DefaultAddressSpace`? The name does suggest it.

kastiglione replied:
partially related: a few of the RemoteAddress constructors in this patch use a literal zero instead of `DefaultAddressSpace`, what's the reason for that?

augusto2112 replied:
> could the second parameter have a default value of swift::remote::RemoteAddress::DefaultAddressSpace?

I was thinking about making it a default constructor, but I think we want to make it explicit that the address space is something that they think about. I'm worried that if someone updates code in the remote inspection library that grabs the address data from a remote address, threads it to some function, and then builds a new RemoteAddress, they may accidentally change the address space. This pattern of RemoteAddress -> StoredPointer -> RemoteAddress is pretty common, so I want to make it harder to make this mistake.

> partially related: a few of the RemoteAddress constructors in this patch use a literal zero instead of DefaultAddressSpace, what's the reason for that?

I just forgot to update those to use the DefaultAddressSpace after I introduced it.

kastiglione replied:
> grabs the address data from a remote address, threads it to some function, and then builds a new RemoteAddress, they may accidentally change the address space. This pattern of RemoteAddress -> StoredPointer -> RemoteAddress is pretty common, so I want to make it harder to make this mistake.

is there a way to provide a safe way of doing this? for example:

```cpp
auto updatedAddr = originalAddr.map([](StoredPointer rawPtr) {
    // do stuff
    return newPtr;
});
```

kastiglione replied:
I think my main issue is the repetitiveness combined with the same value. Despite it being explicit, I could see the default being accidentally used via copy and paste, muscle memory, or LLM generated code.

augusto2112 replied:
I originally had a `fromThis` method on RemoteAddress that would construct a new address and preserve its address space, but Adrian suggested that that was too opaque and not explicit enough. After refactoring, I agree with him.

Another common case that I didn't mention is reading some memory and constructing a new address out of one of its fields. Something like:

```
auto someMetadata = Reader->readSomeMetadata(address);
RemoteAddress newAddress(someMetadata->Field, address.getAddressData());
```

I don't think this is clearer using a map API:

```
auto newAddress = address.map([](StoredPointer) {
  auto someMetadata = Reader->readSomeMetadata(address);
  return someMetata->Field;
};
```

Even if we did adopt such an API, I still think it'd be dangerous to leave the constructor defaulting the address space to the default address space, I think the first thing people would look for when trying to build a new address is its constructor, and they might ignore alternative methods of building the type, and most likely would just write it as:

```
auto someMetadata = Reader->readSomeMetadata(address);
RemoteAddress newAddress(someMetadata->Field);
```

kastiglione commented on lldb/source/Plugins/LanguageRuntime/Swift/LLDBMemoryReader.cpp:
```diff
@@ -455,6 +459,41 @@ bool LLDBMemoryReader::readString(swift::remote::RemoteAddress address,
+                         &pointerSize))
+      return ~0ull;
+
+    if (pointerSize == 4) {
+      uint32_t ptrauthMask32 = 0;
+      if (queryDataLayout(DataLayoutQueryType::DLQ_GetPtrAuthMask, nullptr,
+                          &ptrauthMask32))
+        return (uint64_t)ptrauthMask32;
```
does ptr auth exist for 4 byte pointer systems?

augusto2112 replied:
Yes, apparently they can.

kastiglione replied:
it doesn't have to be part of this PR, but this code makes me think the query API of `DLQ_GetPtrAuthMask` should be always `uint64_t`, regardless of pointer size. It makes code like this more direct, since no call to `DLQ_GetPointerSize` would be needed. For pointer size of 4, the top 4 bytes of the returned value will always be zero and any code that is specific to 32bit can convert as necessary.

augusto2112 committed:
[NFC][lldb] Adapt LLDB to RemoteInspection's RemoteAddress changes

Adapts LLDB to the new field in RemoteAddress. This is an NFC that
maintains the current behavior. A follow up patch will be introduced
later that takes advantage of the new behavior to fix a bug in reading
metadata from files instead of the process.

rdar://148361743

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

kastiglione commented on lldb/source/Plugins/LanguageRuntime/Swift/SwiftMetadataCache.cpp:
```diff
@@ -172,8 +172,9 @@ SwiftMetadataCache::generateHashTableBlob(
-    auto offset = field_descriptor.getAddressData() -
-                  field_descriptors.startAddress().getAddressData();
+    auto offset =
+        field_descriptor.getAddressData().getAddressData() -
+        field_descriptors.startAddress().getAddressData().getAddressData();
```
Are there any other options for the naming of the second `getAddressData()`? `getRawAddress()`?

augusto2112 replied:
I changed `RemoteAddress::getAddressData` to `getRawAddress` and `RemoteRef::getAddressData` to `getRemoteAddress`

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
https://github.com/swiftlang/swift/pull/82638
@swift-ci test

augusto2112 commented:
@swift-ci test
